semantics: support automatic dereference with `.`

amscanne opened the pull request:
Stacked PRs:
 * #4621
 * #4620
 * #4674
 * __->__#4673


--- --- ---

### semantics: support automatic dereference with `.`


Similar to Go, allow for automatic resolution with `.` since we're aware
of whether the type is a pointer or not. This allows us to define `.` as
the canonical syntax for all struct accesses.

Signed-off-by: Adin Scannell <amscanne@meta.com>

jordalgo reviewed (approved):
This could use a few things:
- a changelog entry
- documentation
- at least one runtime test 😄

amscanne commented:
I've updated documentation and tests, and added a changelog entry.

One last thing re: transparency, the intent is to change the behavior of `.` such that it can dereference through pointers. This means that `x.y` and `x->y` are effectively the same thing, and the compiler knows how to treat it based on the type of `x`.

The current thing is weirder IMO, and is what I'm trying to address. We generate an AST that has the deref followed by the field read. But our deref for structs is a lie. I'd love to make this *not* a lie (allow you to actually have structs in bpf memory) but obviously `x->y->z` still needs to be efficient, and it can't be if this is transformed into `(*(*(*x).y).z)`. (We should never implement `->` in this way, so the current "dereference" in the parsed version of `->` has to change either way.)

I'm happy to talk this through, just let me know and we can find a time.

viktormalik commented on tests/parser.cpp:
```diff
@@ -2053,8 +2053,7 @@ TEST(Parser, cast_precedence)
        "   .\n"
-       "    dereference\n"
-       "     builtin: arg0\n"
+       "    builtin: arg0\n"
```
Right, ok, makes sense, no strong objections from my side.

viktormalik commented on docs/tutorial_one_liners.md:
```diff
@@ -5,8 +5,8 @@ This teaches you bpftrace for Linux in 12 easy lessons, where each lesson is a o
 Note: bpftrace 0.19 changed the way probe arguments are accessed (using
-`args.xxx` instead of `args->xxx`). If you are using an older version of
-bpftrace, you will need to use `args->xxx` in the below examples.
+`args.xxx` instead of `args.xxx`). If you are using an older version of
+bpftrace, you will need to use `args.xxx` in the below examples.
```
This should be kept in the original form.

viktormalik commented on tests/runtime/dwarf:
```diff
@@ -31,44 +31,44 @@ TIMEOUT 5
 # Checking backwards compatibility
 NAME uprobe args as pointer
-PROG uprobe:./testprogs/uprobe_test:uprobeFunction1 { printf("c = %c\n", args->c); exit(); }
+PROG uprobe:./testprogs/uprobe_test:uprobeFunction1 { printf("c = %c\n", args.c); exit(); }
```
This one should use `->` since it's explicitly checking for backwards compatibility.

amscanne replied:
Done.

viktormalik commented on tests/runtime/probe:
```diff
@@ -40,7 +40,7 @@ AFTER ./testprogs/syscall read
 # Checking backwards compatibility
 NAME fentry args as a pointer
-PROG fentry:vfs_read { printf("%d\n", args->count); exit(); }
+PROG fentry:vfs_read { printf("%d\n", args.count); exit(); }
```
Same as above.

amscanne replied:
Done.

viktormalik commented on tests/runtime/probe:
```diff
@@ -388,7 +388,7 @@ AFTER ./testprogs/syscall read
 # Checking backwards compatibility
 NAME tracepoint args as pointer
-PROG tracepoint:syscalls:sys_enter_read { printf("%d\n", args->count); exit(); }
+PROG tracepoint:syscalls:sys_enter_read { printf("%d\n", args.count); exit(); }
```
And here.

amscanne replied:
Done.

viktormalik reviewed (commented):
Found a couple of places in the docs/tests where the original `->` should be kept.
